### PR TITLE
Implement export and import of camera settings

### PR DESCRIPTION
**Pull request by adrianusler**
It is now possible to export (/import) camera settings to (/from) an .json file.

**Still to be done:**
- [x] catch exceptions when json file is faulty or damaged [see coderabbit]
- [x] use `tempfile` module instead of deleting file with `Path(...).unlink()` [see coderabbit]

**To be done in future work:** Adapt window size to the specified width/height values?

### CODE REVIEW

**Comment by Michel-Heinz** [comment]
I have merged the PR on the structure customizer. I would like you to use the same interface, where you save the settings in a .json file in the settings/camera directory. You can use the code I wrote in the structure customizer dialog as a reference. Or do you like keeping it a separate format?

**Comment by GereonFeldmann** [comment]
I am of the same opinion as Michel. 
I think we should use the same file format for all settings. 
.json are imo a good option for this. Still .yaml and .toml files would be other options. 
 
I think we should discuss this on the next hackathon.

**Comment by adrianusler** [comment]
I have changed the file format from .npz to .json @Michel-Heinz @GereonFeldmann

File: src/molara/Rendering/camera.py [hunk 1]
```diff
@@ -297,3 +301,66 @@ def calculate_arcball_point(x: float, y: float) -> np.ndarray:
...
+        self.initial_up_vector = pyrr.Vector3(data["initial_up_vector"], dtype=np.float32)
+        self.initial_right_vector = pyrr.Vector3(data["initial_right_vector"], dtype=np.float32)
+        self.target = pyrr.Vector3(data["target"], dtype=np.float32)
+        self.initial_target = pyrr.Vector3(data["initial_target"], dtype=np.float32)
+        self.translation = pyrr.Vector3(data["translation"], dtype=np.float32)
+        self.last_translation = pyrr.Vector3(data["last_translation"], dtype=np.float32)
+        self.rotation = pyrr.Quaternion(data["rotation"])
+        self.last_rotation = pyrr.Quaternion(data["last_rotation"])
+        self.calculate_projection_matrix()
+        self.update()
```
[the remark this answers is not in the document]
Comment (adrianusler) [reply src/molara/Rendering/camera.py:366]: We want to use json, not npz.

File: src/molara/Rendering/camera.py [hunk 2]
```diff
@@ -297,3 +301,66 @@ def calculate_arcball_point(x: float, y: float) -> np.ndarray:
...
+            "initial_target": self.initial_target.tolist(),
+            "last_translation": self.last_translation.tolist(),
+            "rotation": self.rotation.tolist(),
+            "last_rotation": self.last_rotation.tolist(),
+        }
+        if not file_name.endswith(".json"):
+            file_name += ".json"
+        with open(file_name, "w") as file:
+            json.dump(settings, file, indent=4)
+
```
[the remark this answers is not in the document]
Comment (adrianusler) [reply src/molara/Rendering/camera.py:334]: We want to use json, not npz.

File: src/molara/Gui/main_window.py [hunk 3]
```diff
@@ -163,6 +166,30 @@ def export_structure(self) -> None:
...
+        """Export camera settings to .npz file."""
+        file_name = QFileDialog.getSaveFileName(
+            self,
+            "Export camera settings to file",
+            ".",
+            ".json",
+        )[0]
+        if file_name == "":
+            return
+        self.structure_widget.export_camera_settings(file_name)
```
[the remark this answers is not in the document]
Comment (adrianusler) [reply src/molara/Gui/main_window.py:188]: We want to use json, not npz.

File: src/molara/Gui/structure_widget.py [hunk 4]
```diff
@@ -176,6 +176,21 @@
         )
         self.grabFramebuffer().save(filename[0])
 
+    def export_camera_settings(self, filename: str) -> None:
+        """Export camera settings to .npz file.
+
+        :param filename: name of the file to which the camera settings shall be saved
+        """
+        self.camera.export_settings(filename)
```
[the remark this answers is not in the document]
Comment (adrianusler) [reply src/molara/Gui/structure_widget.py:187]: I have opened an issue (#381) for this.

File: src/molara/Gui/structure_widget.py [hunk 5]
```diff
@@ -176,6 +176,21 @@
...
+
+    def import_camera_settings(self, filename: str) -> None:
+        """Import camera settings from .npz file.
+
+        :param filename: name of the file from which the camera settings shall be loaded
+        """
+        self.camera.import_settings(filename)
```
[the remark this answers is not in the document]
Comment (adrianusler) [reply src/molara/Gui/structure_widget.py:194]: I have opened an issue (#381) for this.

File: src/molara/Gui/structure_widget.py [hunk 6]
```diff
@@ -179,6 +179,21 @@
         )
         self.grabFramebuffer().save(filename[0])
 
+    def export_camera_settings(self, filename: str) -> None:
+        """Export camera settings to .npz file.
+
+        :param filename: name of the file to which the camera settings shall be saved
+        """
+        self.camera.export_settings(filename)
+
```
[the remark this answers is not in the document]
Comment (adrianusler) [reply src/molara/Gui/structure_widget.py:188]: Seems excessive to catch exceptions both at the main-window and the structure-widget level. I prefer main window.

File: src/molara/Gui/structure_widget.py [hunk 7]
```diff
@@ -179,6 +179,21 @@
...
+    def import_camera_settings(self, filename: str) -> None:
+        """Import camera settings from .npz file.
+
+        :param filename: name of the file from which the camera settings shall be loaded
+        """
+        self.camera.import_settings(filename)
+        self.update()
+
```
[the remark this answers is not in the document]
Comment (adrianusler) [reply src/molara/Gui/structure_widget.py:196]: Seems excessive to catch exceptions both at the main-window and the structure-widget level. I prefer main window.

File: tests/molara/rendering/test_camera.py [hunk 8]
```diff
@@ -132,6 +135,89 @@ def test_set_distance_from_target(self) -> None:
...
+        assert_array_equal(data["translation"], camera.translation.tolist())
+        assert_array_equal(data["target"], camera.target.tolist())
+        assert_array_equal(data["rotation"], camera.rotation.tolist())
+
+        # reset camera settings
+        self.set_camera_settings_to_zeros()
+
+        # import settings again, then delete json file to clean up
+        camera.import_settings(filename)
+        Path(filename).unlink()
```
[the remark this answers is not in the document]
Comment (adrianusler) [reply tests/molara/rendering/test_camera.py]: See identical comment below.

**Comment by Michel-Heinz** [comment]
Works for me 👍. Can you include the export camera settings to the export that is already there? If you hover ove the export button a new drop down menu should appear with camera and snapshot. I would also suggest moving the import camera setting to the import button, I almost did not find the option. :)

**Comment by adrianusler** [comment]
> Works for me 👍. Can you include the export camera settings to the export that is already there? If you hover ove the export button a new drop down menu should appear with camera and snapshot. I would also suggest moving the import camera setting to the import button, I almost did not find the option. :)

Done @Michel-Heinz.

**Comment by Michel-Heinz** [comment]
top 👍

**Review by Michel-Heinz — APPROVED** [review]

**Pull request merged**